tutor/commands/jobs.py

```diff
@@ -0,0 +1,236 @@
+"""
+Common jobs that must be added both to local, dev and k8s commands.
+"""
+import functools
+import typing as t
+
+import click
+from typing_extensions import ParamSpec
+
+from tutor import config as tutor_config
+from tutor import env, fmt, hooks
+
+
+def add_job_commands(do_command_group: click.Group) -> None:
+    """
+    This is meant to be called with the `local/dev/k8s do` group commands, to add the
+    different `do` subcommands.
+    """
+    subcommands: t.Iterator[click.Command] = hooks.Filters.CLI_DO_COMMANDS.iterate()
+    for subcommand in subcommands:
+        do_command_group.add_command(subcommand)
+
+
+@hooks.Actions.CORE_READY.add()
+def _add_core_init_tasks() -> None:
+    """
+    Declare core init scripts at runtime.
+
+    The context is important, because it allows us to select the init scripts based on
+    the --limit argument.
+    """
+    with hooks.Contexts.APP("mysql").enter():
+        hooks.Filters.COMMANDS_INIT.add_item(("mysql", ("hooks", "mysql", "init")))
+    with hooks.Contexts.APP("lms").enter():
+        hooks.Filters.COMMANDS_INIT.add_item(("lms", ("hooks", "lms", "init")))
+    with hooks.Contexts.APP("cms").enter():
+        hooks.Filters.COMMANDS_INIT.add_item(("cms", ("hooks", "cms", "init")))
+
+
+@click.command("init", help="Initialise all applications")
+@click.option("-l", "--limit", help="Limit initialisation to this service or plugin")
+def initialise(limit: t.Optional[str]) -> t.Iterator[t.Tuple[str, str]]:
+    fmt.echo_info("Initialising all services...")
+    filter_context = hooks.Contexts.APP(limit).name if limit else None
+
+    # Pre-init tasks
+    iter_pre_init_tasks: t.Iterator[
+        t.Tuple[str, t.Iterable[str]]
+    ] = hooks.Filters.COMMANDS_PRE_INIT.iterate(context=filter_context)
+    for service, path in iter_pre_init_tasks:
+        fmt.echo_info(f"Running pre-init task: {'/'.join(path)}")
+        yield service, env.read_template_file(*path)
+
+    # Init tasks
+    iter_init_tasks: t.Iterator[
+        t.Tuple[str, t.Iterable[str]]
+    ] = hooks.Filters.COMMANDS_INIT.iterate(context=filter_context)
+    for service, path in iter_init_tasks:
+        fmt.echo_info(f"Running init task: {'/'.join(path)}")
+        yield service, env.read_template_file(*path)
+
+    fmt.echo_info("All services initialised.")
```

Comment:
Hm, in the interest of reducing the total number of extension points plugin authors need to consider, I was kinda hoping that we could completely implement initialization within the `CLI_DO_COMMANDS` framework, allowing us to deprecate `COMMANDS_INIT` and `COMMANDS_PRE_INIT` and implement `do init` as an ordinary job with no calls to internal Tutor functions.

I understand why it is not currently this way: the `do` framework is designed to allow plugins to *add new* jobs, not to *extend existing* jobs. I don't really have any suggestions here yet... I'll think about it over the weekend.

Reply:
OK, I thought about this more.

I was happy to realize that everything in the new `initialise` function above could basically have also been written in a plugin. There calls to `read_template_file` and `echo_info`, but those are just simple wrappers around standard Python library functions. This is very cool, and it signals to me that `CLI_DO_COMMANDS` is a well-designed interface 👍🏻 

On the other hand, the naming & signature of this corner API feels inconsistent now:
* `CLI_DO_COMMANDS` returns Click commands that yield (service, command template) pairs
* `COMMANDS_INIT` returns (service, command template path) pairs.

The term "command" here is being overloaded between a few different concepts:
1. "command" as in "Click command"
2. "command" as in "a shell command to be run in a container"
3. "command" as in "a (service, shell command) tuple". In some places, we call this a *task*, which I like.

How would you feel about deprecating `COMMANDS_[PRE_]INIT` in favor of new filters:
```python
    #: List of shell commands to be executed during initialization. These commands typically
    #: include database migrations, setting feature flags, etc.
    #:
    #: :parameter list[tuple[str, str]] tasks: list of ``(service, command)`` tasks.
    #:
    #:     - ``service`` is the name of the container in which the task will be executed.
    #:     - ``command`` is a shell command template which will be rendered and executed in ``service``.
    CLI_DO_INIT_TASKS = filters.get("cli:commands:do:init")

    #: ...
    CLI_DO_PRE_INIT_TASKS = filters.get("cli:commands:do:init:pre")
```
(Names up for debate, of course. I wanted to choose names that made it clear that this filter is for adding *tasks* to the existing `do init` command.)

In addition to having different names, these filters have different signatures: they are expected to return `(service, command)` tuples instead of `(service, path)` tuples. This is congruent with the signature of `CLI_DO_COMMANDS`. The mapping from the old filters to the new ones would be pretty straightforward. For example:
```python
    with hooks.Contexts.APP("lms").enter():
        hooks.Filters.COMMANDS_INIT.add_item(("lms", ("hooks", "lms", "init")))
```
into:
```python
    with hooks.Contexts.APP("lms").enter():
        hooks.Filters.CLI_DO_INIT_TASKS.add_item(
            ("lms", env.read_template_file("hooks", "lms", "init"))
        )
```

I realize that this would be a breaking change, and would best be done in a different PR.

Reply:
Yes, your proposition makes a lot of sense. A couple precisions:

- Let's add the `CLI_DO_INIT_TASKS` filter.
- Let's keep the `COMMANDS_INIT` and `COMMANDS_PRE_INIT` filters for now. I don't want to force plugin maintainers to upgrade their plugins right away. We can just print a deprecation warning for the moment.
- We don't even need the `CLI_DO_PRE_INIT_TASKS` filter if we implement filter priorities. This is very easy, based on the existing action priority code.

I'll make these changes right now.